cli/tools/registry/pm/deps.rs

```diff
@@ -0,0 +1,969 @@
+use deno_semver::package::PackageNv;
+use deno_semver::package::PackageReq;
+use deno_semver::package::PackageReqReference;
+use deno_semver::VersionReq;
+use import_map::ImportMap;
+use import_map::ImportMapWithDiagnostics;
+use import_map::SpecifierMapEntry;
+use indexmap::IndexMap;
+use tokio::sync::Semaphore;
+
+use crate::args::CliLockfile;
+use crate::graph_container::MainModuleGraphContainer;
+use crate::graph_container::ModuleGraphContainer;
+use crate::graph_container::ModuleGraphUpdatePermit;
+use crate::jsr::JsrFetchResolver;
+use crate::module_loader::ModuleLoadPreparer;
+use crate::npm::CliNpmResolver;
+use crate::npm::NpmFetchResolver;
+
+use super::ConfigUpdater;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ImportMapKind {
+  Inline,
+  Outline,
+}
+
+#[derive(Clone)]
+pub enum DepLocation {
+  DenoJson(ConfigFileRc, KeyPath, ImportMapKind),
+  PackageJson(PackageJsonRc, KeyPath),
+}
+
+impl DepLocation {
+  pub fn is_deno_json(&self) -> bool {
+    matches!(self, DepLocation::DenoJson(..))
+  }
+
+  pub fn file_path(&self) -> Cow<std::path::Path> {
+    match self {
+      DepLocation::DenoJson(arc, _, _) => {
+        Cow::Owned(arc.specifier.to_file_path().unwrap())
+      }
+      DepLocation::PackageJson(arc, _) => Cow::Borrowed(arc.path.as_ref()),
+    }
+  }
+  fn config_kind(&self) -> super::ConfigKind {
+    match self {
+      DepLocation::DenoJson(_, _, _) => super::ConfigKind::DenoJson,
+      DepLocation::PackageJson(_, _) => super::ConfigKind::PackageJson,
+    }
+  }
+}
+
+struct DebugAdapter<T>(T);
+
+impl<'a> std::fmt::Debug for DebugAdapter<&'a ConfigFileRc> {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    f.debug_struct("ConfigFile")
+      .field("specifier", &self.0.specifier)
+      .finish()
+  }
+}
+impl<'a> std::fmt::Debug for DebugAdapter<&'a PackageJsonRc> {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    f.debug_struct("PackageJson")
+      .field("path", &self.0.path)
+      .finish()
+  }
+}
+
+impl std::fmt::Debug for DepLocation {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    match self {
+      DepLocation::DenoJson(arc, key_path, kind) => {
+        let mut debug = f.debug_tuple("DenoJson");
+        debug
+          .field(&DebugAdapter(arc))
+          .field(key_path)
+          .field(kind)
+          .finish()
+      }
+      DepLocation::PackageJson(arc, key_path) => {
+        let mut debug = f.debug_tuple("PackageJson");
+        debug.field(&DebugAdapter(arc)).field(key_path).finish()
+      }
+    }
+  }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum DepKind {
+  Npm,
+  Jsr,
+}
+
+impl DepKind {
+  pub fn scheme(&self) -> &'static str {
+    match self {
+      DepKind::Npm => "npm",
+      DepKind::Jsr => "jsr",
+    }
+  }
+}
+
+#[derive(Clone, Debug)]
+pub enum KeyPart {
+  Imports,
+  Scopes,
+  Dependencies,
+  DevDependencies,
+  String(String),
```

Comment:
When is this one used?

Reply:
For instance

```
{
  "scopes": {
    "/foo/": {
      "bar": "npm:bar"
    }
}
```

Would have a key path of
```
[KeyPart::Scopes, KeyPart::String("/foo/".into()), KeyPart::String("bar".into())
```

It stores the raw key used to get to the dependency entry